### PR TITLE
Fix query autocomplete with muti-word item and perk names (no more duplicated words)

### PR DESCRIPTION
**Pull request by bjacobel**
Refs https://github.com/DestinyItemManager/DIM/issues/8435

Aims to fix the issue where you type the name of an item that has several words (eg: `arctic haz`) and hit `TAB` and your query is updated to `arctic name:"arctic haze"`. 

This is done by parsing the query up to the cursor location to AST, then traversing the AST from the end backwards to find elements (`keywords`) in the AST that are not part of any filter. Working back from the cursor, these trailing keywords are eliminated from the query if the query contains them in order as bare strings separated by spaces. If the query does not have any unbound keywords or it does not end in a string that matches a joined representation of those keywords, we just fall back on the old behavior of looking for a space with a regex and deleting back to that.

I added a bunch of test cases and also manually tested quite a bit. The only rough edge I've found is that this still falls back to the old behavior if the filter term you're trying to get autocompleted has irregular (duplicate) spaces. I added some notes about how this could be improved. It does work with complex nested queries, negations, the doubled-up `stat:rpm:###` filters, special characters, and all the other edge cases I've thought of. But this'll definitely be something I'd like to get more eyes on to test as I am sure there are kinds of filters I don't even know exist.

### CODE REVIEW

File: src/app/search/autocomplete.test.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ describe('autocompleteTermSuggestions', () => {
     ['(is:blue jun)', 11],
     ['is:bow is:void', 11],
     ['season:>outl', 12],
+    ['not(', 4],
```
Comment (bjacobel) [inline src/app/search/autocomplete.test.ts:18]: This isn't a valid search, it's just a regression test for a bug in here I had at one point.

**Comment by robojumper** [comment]
I like it, especially the bit about traversing the AST. I have some edge cases that would be good to add to your tests at the very least:

* `toil and troub` completes to `toil and name:"toil and trouble"` because the `and` is an operator.
* `rare curio or arctic haz` completes to `rare curio or arctic name:"arctic haze"` because AST traversal collects all four keyword filters.

In general the whole problem is a bit ambiguous, and it's probably impossible to fix both of these at the same time without changing `findLastFilter` to return multiple guesses, though I think the second one could be fixed by only considering direct sibling keyword filters.

(A related idea I had was adding start/end character positions to AST types that the lexer/parser produce and using those, that way you wouldn't have to piece together the string.)

**Comment by delphiactual** [comment]
would not adding `.trim().replace(/s+/,'  ')` to `trailingKeywordArgs` accomplish removing the trailing spaces and the irregular spaces problem?

**Comment by bjacobel** [comment]
@robojumper Thanks for the review. I think `rare curio or arctic haz` is a case where the parser output is a bit odd and that's part of what's giving autocomplete a tough time:

<details>
<summary> parser output for `rare curio or arctic haz` </summary>

```
    {
      "op": "and",
      "operands": [
        {
          "op": "filter",
          "type": "keyword",
          "args": "rare"
        },
        {
          "op": "or",
          "operands": [
            {
              "op": "filter",
              "type": "keyword",
              "args": "curio"
            },
            {
              "op": "filter",
              "type": "keyword",
              "args": "arctic"
            }
          ]
        },
        {
          "op": "filter",
          "type": "keyword",
          "args": "haz"
        }
      ]
    }
```

</details>

The parser output is more expected if `rare curio` is wrapped in quotes, and I fixed that case up and added two more tests for that. Still thinking about how to handle the `toil and trouble` case.

@delphiactual, Good idea - I added that and fixed that case up. Returning column locations of the keyword in the query from the parser would be a better long-term solution for this, but I took a look at that and it seemed quite challenging.

**Comment by bjacobel** [comment]
This should now be ready for another review.

**Review by delphiactual — APPROVED** [review]

File: src/app/search/autocomplete.ts [hunk 2]
```diff
@@ -225,6 +225,51 @@ const lastWordRegex = /(\b[\w:"'<=>]{3,}|#\w*)$/;
...
+      if (keepTraversing && filter.type === 'keyword' && !/\s+/.test(filter.args)) {
+        trailingKeywordArgs.push(filter.args);
+      } else {
+        keepTraversing = false;
+      }
+    },
+    true // traverse in reverse
+  );
+
+  // @TODO: This could be cleaner if the AST parser returned column locations; we wouldn't have to guess at indexes by doing space normalization
```
Comment (bhollis) [inline src/app/search/autocomplete.ts:256]: Shouldn't be too hard to do!

**Review by bhollis — APPROVED** [review]
